### PR TITLE
feat: additions to support using library in two-party protocols

### PR DESCRIPTION
**Pull request by sushisilence**
Changes:
- Add `extract_n_root` method (Used in NI Correct key proof)
- Add minimal versions of SK and PK for serialisation/communication purposes.
- Add serialization for SK, PK using the minimal versions. 
- Fix the `recombine` method. Fixes decrypt_fast failing bug.

### CODE REVIEW

File: crates/sl-paillier/src/lib.rs [hunk 1]
```diff
@@ -32,18 +37,144 @@ pub mod paillier {
...
+    pub fn to_uint(&self) -> Uint<L> {
+        self.0
+    }
+}
+
+#[derive(Debug, PartialEq, Clone)]
 pub struct RawCiphertext<const L: usize>(Uint<L>);
 
-impl<const L: usize> RawCiphertext<L> {
+impl Serialize for RawCiphertext<{ U4096::LIMBS }> {
```
Comment (left-arm) [inline crates/sl-paillier/src/lib.rs:128]: `#[cfg(feature = "serde")]`

**Review by left-arm — CHANGES_REQUESTED** [review]

File: crates/sl-paillier/Cargo.toml [hunk 2]
```diff
@@ -6,9 +6,15 @@ edition = "2021"
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
-crypto-bigint = { version = "0.5.5", default-features = false, features = ["rand_core"] }
+crypto-bigint = { version = "0.5.5", default-features = false, features = [
+  "alloc",
+  "serde",
```
[the remark this answers is not in the document]
Comment (sushisilence) [reply crates/sl-paillier/Cargo.toml:11]: Making optional now

[the remark this answers is not in the document]
Comment (sushisilence) [reply crates/sl-paillier/Cargo.toml]: Updated with changes

File: crates/sl-paillier/src/lib.rs [hunk 3]
```diff
@@ -4,46 +4,173 @@
...
-pub mod paillier {
+#[derive(Debug, Clone, Copy)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub struct MinimalPK<const NN: usize, const N: usize>
+where
+    Uint<N>: Bounded + Encoding,
+    Uint<NN>: Bounded + Encoding,
+{
+    pub n: NonZero<Uint<N>>,
+    pub nn: Uint<NN>,
```
Comment (left-arm) [inline crates/sl-paillier/src/lib.rs:51]: `nn = n*n`

Do we have to  serialize it?

Comment (sushisilence) [reply crates/sl-paillier/src/lib.rs]: We use this value while verifying some proofs. But after check it's okay to remove. Removed in latest commit.

**Review by left-arm — APPROVED** [review]

**Pull request merged**